lib/Driver/Compilation.cpp

```diff
@@ -1164,18 +1149,68 @@ namespace driver {
     collectIncrementalExternallyDependentJobsFromDependencyGraph(
         const bool forRanges) {
       SmallVector<const Job *, 16> ExternallyDependentJobs;
-      // Check all cross-module dependencies as well.
-      forEachOutOfDateIncrementalExternalDependency(
-          forRanges, [&](StringRef dependency) {
-            // If the dependency has been modified since the oldest built file,
-            // or if we can't stat it for some reason (perhaps it's been
-            // deleted?), trigger rebuilds through the dependency graph.
-            for (const Job *marked :
-                 markExternalInDepGraph(dependency, forRanges))
-              ExternallyDependentJobs.push_back(marked);
-          });
+      auto fallbackToExternalBehavior = [&](StringRef external) {
+        for (const auto cmd :
+             markIncrementalExternalInDepGraph(external, forRanges)) {
+          ExternallyDependentJobs.push_back(cmd);
+        }
+      };
```

Comment:
This little lambda could be much clearer. Ideas:  rename `external` to 'externalSwiftModulePath`.
Either add a comment or factor out the for-loop; `markIncrementalExternalInDepGraph` is confusing, because the name implies it marks incremental external <what>? but it is used here to return things. Commands? Are both the side-effect and return values needed? Maybe the name should be something like `findAndMarkExternalCommandsInDependencyGraph`?
Is it idempotent? Maybe there's a great comment I haven't gotten to yet?